presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KDBTree.java

```diff
@@ -0,0 +1,344 @@
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(extent, leafId, left, right);
+        }
+    }
+
+    @JsonCreator
+    public KDBTree(@JsonProperty("rootNode") Node rootNode)
+    {
+        this.rootNode = requireNonNull(rootNode, "rootNode is null");
+    }
+
+    @JsonProperty
+    public Node getRootNode()
+    {
+        return rootNode;
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (obj == null) {
+            return false;
+        }
+
+        if (!(obj instanceof KDBTree)) {
+            return false;
+        }
+
+        KDBTree other = (KDBTree) obj;
+        return this.rootNode.equals(other.rootNode);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(rootNode);
+    }
+
+    public Map<Integer, Rectangle> getLeafNodes()
+    {
+        ImmutableMap.Builder<Integer, Rectangle> leafNodes = ImmutableMap.builder();
+        addLeafNodes(rootNode, leafNodes);
+        return leafNodes.build();
+    }
+
+    private static void addLeafNodes(Node node, ImmutableMap.Builder<Integer, Rectangle> leafNodes)
+    {
+        if (node.leafId.isPresent()) {
+            leafNodes.put(node.leafId.getAsInt(), node.extent);
+        }
+        else {
+            addLeafNodes(node.left.get(), leafNodes);
+            addLeafNodes(node.right.get(), leafNodes);
+        }
+    }
+
+    public Map<Integer, Rectangle> findIntersectingLeafNodes(Rectangle envelope)
+    {
+        ImmutableMap.Builder<Integer, Rectangle> leafNodes = ImmutableMap.builder();
+        addIntersectingLeafNodes(envelope, rootNode, leafNodes);
+        return leafNodes.build();
+    }
+
+    private static void addIntersectingLeafNodes(Rectangle envelope, Node node, ImmutableMap.Builder<Integer, Rectangle> leafNodes)
+    {
+        if (node.extent.disjoint(envelope)) {
+            return;
+        }
+
+        if (node.leafId.isPresent()) {
+            leafNodes.put(node.leafId.getAsInt(), node.extent);
+        }
+        else {
+            addIntersectingLeafNodes(envelope, node.left.get(), leafNodes);
+            addIntersectingLeafNodes(envelope, node.right.get(), leafNodes);
+        }
+    }
+
+    private static final Comparator<Rectangle> BY_X = (first, second) -> ComparisonChain.start()
+            .compare(first.getXMin(), second.getXMin())
+            .compare(first.getYMin(), second.getYMin())
+            .result();
+
+    private static final Comparator<Rectangle> BY_Y = (first, second) -> ComparisonChain.start()
+            .compare(first.getYMin(), second.getYMin())
+            .compare(first.getXMin(), second.getXMin())
+            .result();
+
+    private static final class LeafIdAllocator
+    {
+        private int nextId;
+
+        public int next()
+        {
+            return nextId++;
+        }
+    }
+
+    public static KDBTree buildKDBTree(int maxItemsPerNode, int maxLevels, Rectangle extent, List<Rectangle> items)
```

Comment:
@martint The intent was to put a hard cap on the depth of the tree. I can't remember exactly why I did this. It was necessary when I was using a Quad-tree, but I don't think I need this for the KDB-tree. In case of a Quad-tree, if most of the data was concentrated in one corner of the overall extent, it could take infinite number of iterations to achieve desired partition sizes. Let me remove `maxLevels`.